File: БР1.1/Цветкова Татьяна/labs/lab2/code/shared/src/AppError.ts
export class AppError extends Error {
  public readonly statusCode: number;
  public readonly details?: unknown;

  constructor(message: string, statusCode = 400, details?: unknown) {
    super(message);
    this.statusCode = statusCode;
    this.details = details;
    Error.captureStackTrace?.(this, this.constructor);
  }
}

export class NotFoundError extends AppError {
  constructor(message = "Not found") { super(message, 404); }
}
export class UnauthorizedError extends AppError {
  constructor(message = "Unauthorized") { super(message, 401); }
}
export class ForbiddenError extends AppError {
  constructor(message = "Forbidden") { super(message, 403); }
}
export class ConflictError extends AppError {
  constructor(message = "Conflict") { super(message, 409); }
}
export class ValidationError extends AppError {
  constructor(details: unknown, message = "Validation failed") { super(message, 422, details); }
}

File: БР1.1/Васильев Артур/labs/lr2/vacancy-service/pkg/apperror/apperror.go
package apperror

import "errors"

type Code string

const (
	CodeValidation   Code = "VALIDATION_ERROR"
	CodeUnauthorized Code = "UNAUTHORIZED"
	CodeForbidden    Code = "FORBIDDEN"
	CodeNotFound     Code = "NOT_FOUND"
	CodeConflict     Code = "CONFLICT"
	CodeInternal     Code = "INTERNAL_ERROR"
)

type AppError struct {
	Code    Code
	Message string
	Details []string
	Err     error
}

func (e *AppError) Error() string {
	if e.Err != nil {
		return e.Err.Error()
	}
	return e.Message
}

func (e *AppError) Unwrap() error {
	return e.Err
}

func Validation(message string, details ...string) *AppError {
	return &AppError{Code: CodeValidation, Message: message, Details: details}
}

func Unauthorized(message string) *AppError {
	return &AppError{Code: CodeUnauthorized, Message: message}
}

func Forbidden(message string) *AppError {
	return &AppError{Code: CodeForbidden, Message: message}
}

func NotFound(message string) *AppError {
	return &AppError{Code: CodeNotFound, Message: message}
}

func Conflict(message string) *AppError {
	return &AppError{Code: CodeConflict, Message: message}
}

func Internal(err error) *AppError {
	return &AppError{Code: CodeInternal, Message: "internal server error", Err: err}
}

func IsNotFound(err error) bool {
	var ae *AppError
	return errors.As(err, &ae) && ae.Code == CodeNotFound
}

func IsConflict(err error) bool {
	var ae *AppError
	return errors.As(err, &ae) && ae.Code == CodeConflict
}
